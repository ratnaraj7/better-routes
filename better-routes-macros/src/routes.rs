use quote::{quote, ToTokens};
use syn::parse::{Parse, ParseStream};
use syn::{Ident, ItemStruct, LitStr, Token};

struct Route {
    path: LitStr,
    item_struct: ItemStruct,
    rejection: Option<Ident>,
}

pub struct Routes {
    state: Option<Ident>,
    rejection: Option<Ident>,
    routes: Vec<Route>,
}

impl Parse for Routes {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut state = None;
        let mut rejection = None;
        let mut routes = Vec::new();
        let mut count = 0;

        while !input.is_empty() {
            if count > 0 {
                input.parse::<Token![,]>()?;
                if input.is_empty() {
                    break;
                }
            }

            if input.peek(Ident) {
                let ident: Ident = input.parse()?;
                if ident == "State" {
                    if state.is_some() {
                        return Err(syn::Error::new(ident.span(), "Duplicate State"));
                    }
                    input.parse::<Token![=>]>()?;
                    state = Some(input.parse()?);
                } else if ident == "Rejection" {
                    if rejection.is_some() {
                        return Err(syn::Error::new(ident.span(), "Duplicate Rejection"));
                    }
                    input.parse::<Token![=>]>()?;
                    rejection = Some(input.parse()?);
                }
            } else if input.peek(LitStr) {
                let path: LitStr = input.parse()?;
                input.parse::<Token![=>]>()?;
                let item_struct: ItemStruct = input.parse()?;
                let rejection: Option<Ident> = if input.peek(Token![=>]) {
                    input.parse::<Token![=>]>()?;
                    Some(input.parse()?)
                } else {
                    None
                };
                routes.push(Route {
                    path,
                    item_struct,
                    rejection,
                });
            } else {
                return Err(syn::Error::new(input.span(), "Unexpected Token"));
            }

            count += 1;
        }

        if (state.is_some() || rejection.is_some()) && routes.is_empty() {
            return Err(syn::Error::new(input.span(), "Missing routes"));
        }

        Ok(Routes {
            state,
            rejection,
            routes,
        })
    }
}

impl ToTokens for Routes {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let Routes {
            state,
            rejection: global_rejection,
            routes,
        } = self;

        let typed_paths = routes.iter().map(
            |Route {
                 path,
                 item_struct,
                 rejection,
             }| {
                if let Some(rejection) = rejection.as_ref() {
                    return quote! {
                        #[derive(::axum_extra::routing::TypedPath, ::serde::Deserialize)]
                        #[typed_path(#path, rejection(#rejection))]
                        #item_struct
                    };
                }

                if let Some(rejection) = global_rejection.as_ref() {
                    return quote! {
                        #[derive(::axum_extra::routing::TypedPath, ::serde::Deserialize)]
                        #[typed_path(#path, rejection(#rejection))]
                        #item_struct
                    };
                }

                quote! {
                    #[derive(::axum_extra::routing::TypedPath, ::serde::Deserialize)]
                    #[typed_path(#path)]
                    #item_struct
                }
            },
        );

        let get_all_routes_fn = if let Some(state) = state {
            quote! {
                    fn get_all_routes<T: ::better_routes::MethodHandler<#state>>() -> ::axum::Router<#state> {
                        T::router()
                    }
            }
        } else {
            quote! {
                    fn get_all_routes<T: ::better_routes::MethodHandler>() -> ::axum::Router {
                        T::router()
                    }
            }
        };

        let mergers = routes.iter().map(|Route { item_struct, .. }| {
            let ident = &item_struct.ident;
            quote! {
                app = app.merge(get_all_routes::<#ident>());
            }
        });

        let router_fn = if let Some(state) = state {
            quote! {
                pub fn router() -> ::axum::Router<#state> {
                    let mut app = ::axum::Router::new();
                    #(#mergers)*
                    app
                }
            }
        } else {
            quote! {
                pub fn router() -> ::axum::Router {
                    let mut app = ::axum::Router::new();
                    #(#mergers)*
                    app
                }
            }
        };

        tokens.extend(typed_paths);
        tokens.extend(get_all_routes_fn);
        tokens.extend(router_fn);
    }
}
