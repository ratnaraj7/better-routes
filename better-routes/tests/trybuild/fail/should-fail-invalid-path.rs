use axum_extra::routing::RouterExt;
use better_routes::routes;
use serde::Deserialize;
#[derive(Deserialize)]
struct Home;
#[derive(Deserialize)]
struct About;
async fn about(_: About) {}
routes! {
    name => AllRoutes,
    "/" => Home {
        get => about,
    },
    "/about" => About {
        get => about
    },
}
fn main() {}
