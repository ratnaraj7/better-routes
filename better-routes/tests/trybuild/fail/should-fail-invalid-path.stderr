error[E0277]: the trait bound `(axum_core::extract::private::ViaParts, About): SecondElementIs<Home>` is not satisfied
  --> tests/trybuild/fail/should-fail-invalid-path.rs:12:16
   |
12 |         get => about,
   |                ^^^^^ the trait `SecondElementIs<Home>` is not implemented for `(axum_core::extract::private::ViaParts, About)`
   |
   = help: the following other types implement trait `SecondElementIs<P>`:
             (M, P)
             (M, P, T1)
             (M, P, T1, T2)
             (M, P, T1, T2, T3)
             (M, P, T1, T2, T3, T4)
             (M, P, T1, T2, T3, T4, T5)
             (M, P, T1, T2, T3, T4, T5, T6)
             (M, P, T1, T2, T3, T4, T5, T6, T7)
           and $N others
note: required by a bound in `typed_get`
  --> $CARGO/axum-extra-0.9.3/src/routing/mod.rs
   |
   |     fn typed_get<H, T, P>(self, handler: H) -> Self
   |        --------- required by a bound in this associated function
...
   |         T: SecondElementIs<P> + 'static,
   |            ^^^^^^^^^^^^^^^^^^ required by this bound in `RouterExt::typed_get`
